timer

matthewzorin opened the pull request:
Merge solution for the third lesson homework. May need to add README.md file

matthewzorin committed:
timer: Make init module

	Add sysfs class, read attribute
	Add Makefile and format file

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Add seconds counter

        Add test script and log file

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Add high resolution timer

	Add simple script for checking the high resolution timer state
	Add sysfs read interface

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Change high resolution timer show mode to float point

	Remove useless callback function and its adding to hrtimer structure

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

yekovalyov commented on timer/timer.c:
```diff
@@ -37,13 +38,19 @@ static ssize_t inter_show(struct class *class, struct class_attribute *attr,
+		}
+		hr_sec[len - DOT_POS] = '.';
+	}
+	sprintf(buffer, "time = %s s\n", hr_sec);
```
What is the reason for timer printing redesign?

matthewzorin replied:
ktime_t store value in nanoseconds. So, for getting time string in seconds, I have used this string conversion. Before I was getting the same via division and modulo operations

matthewzorin replied:
I have two questions about the implementation of boot option (via DTS). What kind of driver data structure (like "platform_driver" or "i2c_driver") should I use? And which .dtsi and .dts files are proper for changing?

yekovalyov replied:
Can you simplify this code? I think previous implementation was better.

matthewzorin committed:
timer: Change sysfs show mode

	Replace the time value conversion from kernel module to bash script
	Remove plain text from showing in sysfs interfaces

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Change absolute time interface

	Show absolute time of previous reading (before was current)

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

yekovalyov commented on timer/timer_test.sh:
```diff
@@ -0,0 +1,14 @@
+done
+make uninstall
+
+exit $RETURN_SUCCESS
```
What is the value of RETURN_SUCCESS?
Is this exit required?

matthewzorin replied:
No, it's just code style in this example. exit $RETURN_SUCCESS directly indicates the end of script

yekovalyov replied:
I suppose that using uninitialized variables is not good practice.

matthewzorin replied:
I've removed this unnecessary exit call.

yekovalyov commented on timer/timer.c:
```diff
@@ -11,6 +11,7 @@
 #define DOT_POS 9
```
Is this define used after code change?

matthewzorin replied:
No, it's useless. I forgot to remove this define making changes

matthewzorin committed:
timer: Remove useless define

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Refactor code

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>

matthewzorin committed:
timer: Remove exit command from each test script

Signed-off-by: Zorin Matvii <zorinmatvej87@gmail.com>
